From 7fe2474e749f9fb6247c6bcb4167c8d27f1dcc7c Mon Sep 17 00:00:00 2001
From: Vincent Abriou <vincent.abriou@st.com>
Date: Mon, 23 Jan 2017 13:32:34 +0100
Subject: [PATCH 1/4] compositor-st: don't repaint  primary plane if no damage

If we don't have any damage for the primary plane, then don't force a
repaint; simply reuse the old buffer we already have.

backport of https://patchwork.freedesktop.org/patch/121776/

Change-Id: Ic966ef92b7b55fd3c892b311956098e9619c602e
Signed-off-by: Vincent Abriou <vincent.abriou@st.com>
Reviewed-on: https://gerrit.st.com/63826
Reviewed-by: Fabien DESSENNE <fabien.dessenne@st.com>
Tested-by: Fabien DESSENNE <fabien.dessenne@st.com>
---
 libweston/compositor-st.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/libweston/compositor-st.c b/libweston/compositor-st.c
index 25f6893..ea9f27c 100644
--- a/libweston/compositor-st.c
+++ b/libweston/compositor-st.c
@@ -757,7 +757,12 @@ drm_output_render(struct drm_output *output, pixman_region32_t *damage)
 	struct weston_compositor *c = output->base.compositor;
 	struct drm_backend *b = to_drm_backend(c);
 
-	if (b->use_pixman)
+	if (!pixman_region32_not_empty(damage) && output->current) {
+		if (!output->next)
+			output->next = output->current;
+		weston_xlog("no dammage on primary plane, do not repaint it\n");
+		return;
+	} else if (b->use_pixman)
 		drm_output_render_pixman(output, damage);
 	else
 		drm_output_render_gl(output, damage);
@@ -1082,7 +1087,8 @@ page_flip_handler(int fd, unsigned int frame,
 	 * we just want to page flip to the current buffer to get an accurate
 	 * timestamp */
 	if (output->page_flip_pending) {
-		drm_output_release_fb(output, output->current);
+		if (output->current != output->next)
+			drm_output_release_fb(output, output->current);
 		output->current = output->next;
 		output->next = NULL;
 	}
-- 
2.7.4

